custom_metrics/fugu-apis.js

```diff
@@ -0,0 +1,213 @@
+    regEx: /new IdleDetector\(/g,
+    where: "JavaScript",
+  },
+  "Storage Foundation": {
+    regEx: /storageFoundation\.open\(/g,
+    where: "JavaScript",
+  },
+  "Handwriting Recognition": {
+    regEx: /navigator\.queryHandwritingRecognizerSupport\(/g,
+    where: "JavaScript",
+  },
+  "Compute Pressure": {
+    regEx: /new ComputePressureObserver\(/g,
+    where: "JavaScript",
+  },
+  "Ambient Light Sensor": {
+    regEx: /new AmbientLightSensor\(\)/g,
+    where: "JavaScript",
+  },
+  "File Handling": {
+    regEx: /launchQueue\.setConsumer\(/g,
+    where: "JavaScript",
+  },
+  "Notification Triggers": {
+    regEx: /showTrigger\:/g,
+    where: "JavaScript",
+  },
+  "Local Font Access": {
+    regEx: /navigator\.fonts\.query\(\);/g,
+    where: "JavaScript",
+  },
+  "Multi-Screen Window Placement": {
+    regEx: /getScreens\(\)/g,
+    where: "JavaScript",
+  },
+  "WebSocketStream": {
+    regEx: /new WebSocketStream\(/g,
+    where: "JavaScript",
+  },
+  "WebTransport": {
+    regEx: /new WebTransport\(/g,
+    where: "JavaScript",
+  },
+  "Web Share Target": {
+    regEx: /"share_target"\:/g,
+    where: "Web App Manifest",
+  },
+  "Web Share Target (Files)": {
+    regEx: /"enctype"\: "multipart\/form\-data"/g,
+    where: "Web App Manifest",
+  },
+  "Shortcuts": {
+    regEx: /"shortcuts"\:/g,
+    where: "Web App Manifest",
+  },
+  "Declarative Link Capturing": {
+    regEx: /"capture_links"\:/g,
+    where: "Web App Manifest",
+  },
+  "Tabbed Application Mode": {
+    regEx: /"tabbed"/g,
+    where: "Web App Manifest",
+  },
+  "Window Controls Overlay": {
+    regEx: /"window\-controls\-overlay"/g,
+    where: "Web App Manifest",
+  },
+  "URL Handlers": {
+    regEx: /"url_handlers"/g,
+    where: "Web App Manifest",
+  },
+  "Protocol Handlers": {
+    regEx: /"protocol_handlers"/g,
+    where: "Web App Manifest",
+  },
+};
+
+// To make sure we don't match on, e.g., blog posts that contain the patterns,
+// make sure that the file names fulfill certain conditions as a heuristic.
+const checkURLConditions = (where, url) => {
+  // If the pattern has to occur in JavaScript, make sure the file name
+  // includes either `.js` or `.mjs`.
+  if (where === "JavaScript" && /\.m?js/.test(url)) {
+    return true;
+    // If the pattern has to occur in the Web App Manifest, make sure the file
+    // name includes either `.json` or `.webmanifest`.
+  } else if (
+    where === "Web App Manifest" &&
+    (/\.webmanifest/.test(url) || /\.json/.test(url))
+  ) {
+    return true;
+  }
+  return false;
+};
+
+// Iterate over all response bodies and over all patterns and populate the
+// result object.
+const result = {};
+responseBodies.forEach((har) => {
+  for (const [key, value] of Object.entries(patterns)) {
+    if (value.regEx.test(har.response_body)) {
+      if (result[har.url] && !result[har.url].includes(key)) {
+        if (checkURLConditions(value.where, har.url)) {
+          result[har.url].push(key);
```

Comment:
WDYT about flipping this so that the result key is the feature name and the array of values are the URLs?

For the analysis I can imagine the most important thing being "does this page use feature X" and that may be more straightforward to extract from the JSON by checking if the key exists in the object rather than scanning all URLs' arrays.

Reply:
Sounds good.